xbmc/favourites/ContextMenus.cpp

```diff
@@ -8,73 +8,202 @@
 {
-  bool CFavouriteContextMenuAction::IsVisible(const CFileItem& item) const
-  {
-    return URIUtils::IsProtocol(item.GetPath(), "favourites");
-  }
+  return URIUtils::IsProtocol(item.GetPath(), "favourites");
+}
 
-  bool CFavouriteContextMenuAction::Execute(const std::shared_ptr<CFileItem>& item) const
+bool CFavouriteContextMenuAction::Execute(const std::shared_ptr<CFileItem>& item) const
+{
+  CFileItemList items;
+  CServiceBroker::GetFavouritesService().GetAll(items);
+  for (const auto& favourite : items)
   {
-    CFileItemList items;
-    CServiceBroker::GetFavouritesService().GetAll(items);
-    for (const auto& favourite : items)
+    if (favourite->GetPath() == item->GetPath())
     {
-      if (favourite->GetPath() == item->GetPath())
-      {
-        if (DoExecute(items, favourite))
-          return CServiceBroker::GetFavouritesService().Save(items);
-      }
+      if (DoExecute(items, favourite))
+        return CServiceBroker::GetFavouritesService().Save(items);
     }
-    return false;
   }
+  return false;
+}
+
+bool CMoveUpFavourite::DoExecute(CFileItemList& items, const std::shared_ptr<CFileItem>& item) const
+{
+  return FAVOURITES_UTILS::MoveItem(items, item, -1);
+}
+
+bool CMoveUpFavourite::IsVisible(const CFileItem& item) const
+{
+  return CFavouriteContextMenuAction::IsVisible(item) && FAVOURITES_UTILS::ShouldEnableMoveItems();
+}
 
-  bool CMoveUpFavourite::DoExecute(CFileItemList& items,
+bool CMoveDownFavourite::DoExecute(CFileItemList& items,
                                    const std::shared_ptr<CFileItem>& item) const
-  {
-    return FAVOURITES_UTILS::MoveItem(items, item, -1);
-  }
+{
+  return FAVOURITES_UTILS::MoveItem(items, item, +1);
+}
 
-  bool CMoveUpFavourite::IsVisible(const CFileItem& item) const
-  {
-    return CFavouriteContextMenuAction::IsVisible(item) &&
-           FAVOURITES_UTILS::ShouldEnableMoveItems();
-  }
+bool CMoveDownFavourite::IsVisible(const CFileItem& item) const
+{
+  return CFavouriteContextMenuAction::IsVisible(item) && FAVOURITES_UTILS::ShouldEnableMoveItems();
+}
 
-  bool CMoveDownFavourite::DoExecute(CFileItemList& items,
-                                     const std::shared_ptr<CFileItem>& item) const
-  {
-    return FAVOURITES_UTILS::MoveItem(items, item, +1);
-  }
+bool CRemoveFavourite::DoExecute(CFileItemList& items, const std::shared_ptr<CFileItem>& item) const
+{
+  return FAVOURITES_UTILS::RemoveItem(items, item);
+}
+
+bool CRenameFavourite::DoExecute(CFileItemList&, const std::shared_ptr<CFileItem>& item) const
+{
+  return FAVOURITES_UTILS::ChooseAndSetNewName(*item);
+}
+
+bool CChooseThumbnailForFavourite::DoExecute(CFileItemList&,
+                                             const std::shared_ptr<CFileItem>& item) const
+{
+  return FAVOURITES_UTILS::ChooseAndSetNewThumbnail(*item);
+}
 
-  bool CMoveDownFavourite::IsVisible(const CFileItem& item) const
+namespace
+{
+std::shared_ptr<CFileItem> ResolveFavouriteItem(const CFileItem& item)
```

Comment:
Do these items have any sort of vfs prefix (e.g. `favourite://`)? If so maybe the item resolution should be moved to IDIrectory::Resolve() ?

Reply:
There is no such thing like Resolve in IDirectory. How would this look like? But honestly I don't feel like reworking all this without really good reason.

Reply:
There is, see https://github.com/search?q=repo%3Axbmc%2Fxbmc%20Resolve(CFileItem%26%20item)%20const&type=code

The reason is to try to reuse the existing interface instead of creating a new one :)

Reply:
Oh, my bad. And how would the using code look like here? Need some guidance...